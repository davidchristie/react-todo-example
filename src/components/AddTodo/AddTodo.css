.AddTodo {
  padding: 5px;
}

.AddTodo input {
  font-size: larger;
  padding: 10px;
  width: 80%;
}
