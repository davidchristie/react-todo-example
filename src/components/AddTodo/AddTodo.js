import PropTypes from 'prop-types'
import React from 'react'

import './AddTodo.css'

class AddTodo extends React.Component {
  constructor (props) {
    super(props)
    this._onChange = this._onChange.bind(this)
    this._onSubmit = this._onSubmit.bind(this)
    this.state = {
      text: ''
    }
  }

  _isValidTodo () {
    return this.state.text.length > 0
  }

  _onChange (event) {
    this.setState({
      text: event.target.value
    })
  }

  _onSubmit (event) {
    event.preventDefault()
    if (this._isValidTodo()) {
      if (this.props.onSubmit) {
        const newItem = {
          completed: false,
          text: this.state.text
        }
        this.props.onSubmit(newItem)
      }
      this.setState({
        text: ''
      })
    }
  }

  render () {
    return (
      <div className='AddTodo'>
        <form onSubmit={this._onSubmit}>
          <input
            onChange={this._onChange}
            placeholder='Add an item...'
            type='text'
            value={this.state.text}
          />
        </form>
      </div>
    )
  }
}

AddTodo.propTypes = {
  onSubmit: PropTypes.func
}

export default AddTodo
