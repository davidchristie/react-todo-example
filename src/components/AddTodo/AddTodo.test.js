import { mount, shallow } from 'enzyme'
import React from 'react'

import AddTodo from './AddTodo'

it('renders correctly', () => {
  expect(shallow(<AddTodo />)).toMatchSnapshot()
})

it('prevents the default submit action', () => {
  const event = {
    preventDefault: jest.fn()
  }
  const wrapper = mount(<AddTodo />)
  wrapper.find('input').simulate('submit', event)
  expect(event.preventDefault.mock.calls.length).toBe(1)
})

it(`doesn't submit anything if text is empty`, () => {
  const props = {
    onSubmit: jest.fn()
  }
  const wrapper = mount(<AddTodo {...props} />)
  wrapper.find('input').simulate('submit', {
    preventDefault: () => {}
  })
  expect(props.onSubmit.mock.calls.length).toBe(0)
})

it('clears the text after submitting', () => {
  const wrapper = mount(<AddTodo />)
  const input = wrapper.find('input')
  input.simulate('change', {target: {value: 'Some text'}})
  input.simulate('submit', {
    preventDefault: () => {}
  })
  expect(wrapper.state().text).toBe('')
})

it('submits a new todo item if input is valid', () => {
  const text = 'Test todo item text'
  const props = {
    onSubmit: jest.fn()
  }
  const wrapper = mount(<AddTodo {...props} />)
  const input = wrapper.find('input')
  input.simulate('change', {target: {value: text}})
  input.simulate('submit', {
    preventDefault: () => {}
  })
  expect(props.onSubmit.mock.calls[0]).toEqual([
    {
      completed: false,
      text
    }
  ])
})
