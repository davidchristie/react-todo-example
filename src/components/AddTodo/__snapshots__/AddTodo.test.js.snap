// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="AddTodo"
>
  <form
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      placeholder="Add an item..."
      type="text"
      value=""
    />
  </form>
</div>
`;
